cabal-install/src/Distribution/Client/CmdStatus.hs

```diff
@@ -0,0 +1,381 @@
+  }
+  deriving (Show, Read, Eq, Ord)
+
+data CompilerInformation = CompilerInformation
+  { ciFlavour :: CompilerFlavor
+  , ciCompilerId :: CompilerId
+  , ciPath :: FilePath
+  }
+  deriving (Show, Read, Eq, Ord)
+
+data ResolvedTarget = ResolvedTarget
+  { rtOriginalTarget :: String
+  -- | UnitId of the resolved target.
+  -- If 'Nothing', then the given target can not be resolved
+  -- to a target in this project.
+  , rtUnitId :: Maybe UnitId
+  }
+  deriving (Show, Read, Eq, Ord)
+
+serialise :: Verbosity -> Flag StatusOutputFormat -> StatusInformation -> IO String
+serialise verbosity NoFlag      _  =
+  die' verbosity $ "Could not serialise Status information. "
+                ++ "The flag '--output-format' is required."
+
+serialise _         (Flag JSON) si = pure $ Json.encodeToString $ Json.object $
+  [ "cabal-version" .= jdisplay (siCabalVersion si)
+  ]
+  ++ prettyCompilerInfo (siCompiler si)
+  ++ prettyTargetResolving (siTargetResolving si)
+  where
+    prettyCompilerInfo Nothing = []
+    prettyCompilerInfo (Just ci) =
+      [ "compiler" .= Json.object
+        [ "flavour"     .= jdisplay (ciFlavour ci)
+        , "compiler-id" .= jdisplay (ciCompilerId ci)
+        , "path"        .= Json.String (ciPath ci)
+        ]
+      ]
+
+    prettyTargetResolving Nothing = []
+    prettyTargetResolving (Just rts) =
+      [ "targets" .= Json.Array (fmap prettyResolvedTarget rts)
+      ]
+      where
+        prettyResolvedTarget rt = Json.object
+          [ "target" .= Json.String (rtOriginalTarget rt)
+          , "unit-id" .= maybe Json.Null jdisplay (rtUnitId rt)
+          ]
+
+
+-- ----------------------------------------------------------------------------
+-- Helpers for determining and serialising compiler information
+-- ----------------------------------------------------------------------------
+
+requireCompilerProg :: Verbosity -> Compiler -> IO Program
+requireCompilerProg verbosity compiler =
+  case compilerFlavor compiler of
+    GHC -> pure ghcProgram
+    GHCJS -> pure ghcjsProgram
+    flavour -> die' verbosity $
+      "status: Unsupported compiler flavour: "
+      <> prettyShow flavour
+
+mkCompilerInfo :: ConfiguredProgram -> Compiler -> CompilerInformation
+mkCompilerInfo compilerProgram compiler =
+  CompilerInformation (compilerFlavor compiler) (compilerId compiler) (programPath compilerProgram)
+
+-- ----------------------------------------------------------------------------
+-- Helpers for determining and serialising the unit-id
+-- ----------------------------------------------------------------------------
+
+mkBuildInfoJson :: ElaboratedInstallPlan -> TargetsMap -> Map TargetSelector String -> [String] -> [ResolvedTarget]
+mkBuildInfoJson elaboratedPlan targetsMap tsMap unresolvableTargetStrings =
+  [ ResolvedTarget str (Just uid)
+  | uid <- Map.keys subsetInstallPlan
+  , (_, tss) <- targetsMap Map.! uid
+  , str <- ordNub $ map tsToOriginalTarget $ toList tss
+  ]
```

Comment:
I've no idea what this does or if it's correct :D

Reply:
Tried to improve this piece of code, but it needs some refactoring anyway, due to its usage of `Map.!` which yields sub-par error messages.